### PR TITLE
Added ability to update GitHub gists

### PR DESCRIPTION
**Pull request by cshashwat**
Added ability to update GitHub gists

## Description

A user can now update gists in GitHub, if he/she owns the gist. Previously, every time a user clicked "Public new gist" or "Private new gist" under the "Share" menu, a new copy was created. The proposed feature would allow users to update their gists in place, with GitHub showing each new version.

If a user does not own the gist (e.g. importing the gist from another person's account), he/she will not be able to update the gist. However upon sharing as a public or private gist, the user now has permission and can update his/her own copy of the gist.

The "Share" menu is updated to reflect the new "Update Existing Gist" option; strings have been updated for clarity for preexisting options as well. If a user imports a gist, then deletes it from GitHub and tries to click "Update" from Script Lab, the app will trigger a full refresh, similar to the existing import failure. Publishing a gist, public or private, does not create new local copies of that gist, but updates the gist's ID instead with the new one provided by GitHub.

## Related Issue

https://github.com/OfficeDev/script-lab/issues/27

## Motivation and Context

See above link to issue. Users may or may not want additional copies of gists in their accounts; they now have the functionality to choose.

## How Has This Been Tested?

Testing was done by following the existing scenarios in the new file 'TESTING.md' under the 'Share' section, with other scenarios added to reflect the new proposed feature.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by WrathOfZombies** [comment]
@cshashwat Will take a look at it by tomorrow EOD. Thanks for the contribution 👍

File: TESTING.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  * Running deleted snippet has reasonable behavior.
+  * [Side-by-side runner]:
+    * Edit to code causes runner want to refresh.
+    * Whether starting from error or going to error state and back out, should act correctly.
+  * Note: when running a script on localhost, you may see two compiler warnings for a critical dependency in the console log. These two warnings always appear in debug mode and are normal.
+* Sharing:
+  * Copying to clipboard works
+  * Can share as gist, public and private
+  * Can update existing gist
+  * Deleting a gist and then trying to update it via local copy should prompt full refresh, after which update option should be gone
```
Comment (dsaunders45459) [inline TESTING.md]: Should mention the case that the "Update existing Gist" button becomes available immediately after initial publish of a new Gist.

File: TESTING.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  * Erroneous code (e.g, syntax error) shows error correctly.
+  * Running deleted snippet has reasonable behavior.
+  * [Side-by-side runner]:
+    * Edit to code causes runner want to refresh.
+    * Whether starting from error or going to error state and back out, should act correctly.
+  * Note: when running a script on localhost, you may see two compiler warnings for a critical dependency in the console log. These two warnings always appear in debug mode and are normal.
+* Sharing:
+  * Copying to clipboard works
+  * Can share as gist, public and private
+  * Can update existing gist
```
Comment (dsaunders45459) [inline TESTING.md]: Should also mention to test the case where you import a Gist that you do not own and try to update it.

File: src/client/app/containers/app.ts [hunk 3]
```diff
@@ -144,7 +145,7 @@ export class AppComponent {
         this._store.dispatch(new UI.ToggleImportAction(true));
     }
 
-    shareGist(isPublic: boolean) {
+    shareGist(isPublic: boolean, update: boolean) {
```
Comment (dsaunders45459) [inline src/client/app/containers/app.ts]: The project should follow and document a uniform naming convention. It looks like booleans should be prefixed with is* for this function.

File: src/client/app/effects/snippet.ts [hunk 4]
```diff
@@ -144,6 +144,29 @@ export class SnippetEffects {
...
+        .ofType(Snippet.SnippetActionTypes.UPDATE_INFO)
+        .map(( { payload } ) => {
+            let { id, name, description, gist, owned } = payload;
+            let snippet: ISnippet = storage.lastOpened;
+            if (storage.snippets.contains(id)) {
+                snippet = storage.snippets.get(id);
+
+                /* check if fields are undefined or null */
+                name != null ? snippet.name = name : snippet.name = snippet.name;
+                description != null ? description = payload.description : snippet.description = snippet.description;
```
Comment (dsaunders45459) [inline src/client/app/effects/snippet.ts]: I am not sure why we assign to description and then no longer use the local variable.

File: src/client/app/effects/snippet.ts [hunk 5]
```diff
@@ -144,6 +144,29 @@ export class SnippetEffects {
...
+    updateInfo$: Observable<Action> = this.actions$
+        .ofType(Snippet.SnippetActionTypes.UPDATE_INFO)
+        .map(( { payload } ) => {
+            let { id, name, description, gist, owned } = payload;
+            let snippet: ISnippet = storage.lastOpened;
+            if (storage.snippets.contains(id)) {
+                snippet = storage.snippets.get(id);
+
+                /* check if fields are undefined or null */
+                name != null ? snippet.name = name : snippet.name = snippet.name;
```
Comment (dsaunders45459) [inline src/client/app/effects/snippet.ts]: snippet.name = snippet.name looks like a complicated no-op. Can we just change all of these
if (name != null) {
    snippet.name = name;
}

File: src/client/app/reducers/github.ts [hunk 6]
```diff
@@ -48,7 +48,8 @@ export function reducer(state = initialState, action: GitHubActions): GitHubStat
         }
 
         case GitHubActionTypes.SHARE_PRIVATE_GIST:
-        case GitHubActionTypes.SHARE_PUBLIC_GIST: {
+        case GitHubActionTypes.SHARE_PUBLIC_GIST:
+        case GitHubActionTypes.UPDATE_GIST: {
             AI.trackEvent(action.type);
             return { ...state, sharing: true };
```
Comment (dsaunders45459) [inline src/client/app/reducers/github.ts]: Can we log a hashed gist id?

File: src/client/app/services/github.service.ts [hunk 7]
```diff
@@ -95,7 +95,7 @@ export class GitHubService {
         if (this.profile == null) {
             return Observable.of([]);
         }
-        let url = `${this._baseUrl}/users/${this.profile.login}/gists`;
+        let url = `${this._baseUrl}/users/${this.profile.login}/gists?timestamp=${(new Date).getTime()}`;
```
Comment (dsaunders45459) [inline src/client/app/services/github.service.ts]: It looks like angular doesn't support cache busting by default like jQuery, but you could also add  'Cache-Control' : 'no-cache' to the headers to prevent the browser cache.

File: src/interfaces/playground.d.ts [hunk 8]
```diff
@@ -6,6 +6,7 @@ interface ITemplate {
     ///////////////////////////////////////////////////////////////////////////////////////////////////
     id?: string;
     gist?: string;
+    owned?: boolean;
```
Comment (dsaunders45459) [inline src/interfaces/playground.d.ts]: We should follow standard naming conventions throughout the project. I think this should be named isOwned

File: TESTING.md (hunk 1 above)
Comment (cshashwat) [inline TESTING.md]: Added.

File: TESTING.md (hunk 2 above)
Comment (cshashwat) [inline TESTING.md]: Added.

File: src/client/app/containers/app.ts (hunk 3 above)
Comment (cshashwat) [inline src/client/app/containers/app.ts]: Added.

File: src/client/app/effects/snippet.ts (hunk 4 above)
Comment (cshashwat) [inline src/client/app/effects/snippet.ts]: This was a typo on my part, thanks for catching it. Fixed this and modified the ternary operators to be if statements instead, since the 'snippet.field = snippet.field' is an unnecessary no-op.

File: src/client/app/effects/snippet.ts (hunk 5 above)
Comment (cshashwat) [inline src/client/app/effects/snippet.ts]: Added; see above comment.

File: src/client/app/reducers/github.ts (hunk 6 above)
Comment (cshashwat) [inline src/client/app/reducers/github.ts]: Added under UPDATE_INFO in 'snippet.ts' in 'reducers'; it logs the gist id using the same sha hash function.

File: src/client/app/services/github.service.ts (hunk 7 above)
Comment (cshashwat) [inline src/client/app/services/github.service.ts]: Discussed offline. Added optional parameter for get() function in 'app/services/request.ts' to use cache or not; if this is set to 'false', then the timestamp random number param is applied. The GitHub API seems to reject using 'Cache-Control' in the headers. Would welcome other suggestions.

File: src/interfaces/playground.d.ts (hunk 8 above)
Comment (cshashwat) [inline src/interfaces/playground.d.ts]: Changed throughout code to 'isOwned'

File: src/client/app/reducers/snippet.ts [hunk 9]
```diff
@@ -81,6 +82,16 @@ export function reducer(state = initialState, action: SnippetActions | GitHubAct
         case SnippetActionTypes.STORE_UPDATED:
             return { ...state, loading: false };
 
+        case SnippetActionTypes.UPDATE_INFO: {
+            let updatedInfo = { };
+            if (action.payload.gist != null) {
+                updatedInfo['gistHashedId'] = sha1(action.payload.gist.toString());
```
Comment (dsaunders45459) [inline src/client/app/reducers/snippet.ts]: Is gist already a string? If it is there is no need to call toString on it.

File: src/client/app/helpers/snippet.helper.ts [hunk 10]
```diff
@@ -72,7 +73,6 @@ export const snippetFieldSortingOrder: { [key: string]: number } = {
 export function getSnippetDefaults(): ISnippet {
     return {
         id: '',
-        gist: '',
```
Comment (Zlatkovsky) [inline src/client/app/helpers/snippet.helper.ts]: Why did you remove the empty-string gist?  It seemed to be a reasonable default...

File: src/client/app/helpers/strings.ts [hunk 11]
```diff
@@ -52,7 +53,8 @@ export class Strings {
     static readonly githubLogoutFailed = 'Failed to sign out of GitHub';
     static readonly profileCheckFailed = 'Failed to get the GitHub profile';
     static readonly gistRetrieveFailed = 'Failed to retrieve GitHub gists';
-    static readonly gistShareFailed = 'Failed to share the GitHub gist';
+    static readonly gistShareFailedBody = 'Failed to share the GitHub gist';
+    static readonly gistShareFailedTitle = 'Share failed';
```
Comment (Zlatkovsky) [inline src/client/app/helpers/strings.ts]: I'd probably do "Sharing failed" instead of "Share failed".
In general, we've been doing string peer-review by putting `/* NEEDS STRING REVIEW */` next to new strings, and having Michael Saunders then look through them.  Search for this phrase and see other instances where I have it

File: src/client/app/services/github.service.ts [hunk 12]
```diff
@@ -96,7 +96,7 @@ export class GitHubService {
             return Observable.of([]);
         }
         let url = `${this._baseUrl}/users/${this.profile.login}/gists`;
-        return this._request.get<IGist[]>(url, ResponseTypes.JSON, this._headers);
+        return this._request.get<IGist[]>(url, ResponseTypes.JSON, this._headers, false);
```
Comment (Zlatkovsky) [inline src/client/app/services/github.service.ts]: Add a parameter comment:

    return this._request.get<IGist[]>(url, ResponseTypes.JSON, this._headers, false /*useCache*/);

File: src/client/app/services/request.ts [hunk 13]
```diff
@@ -19,7 +19,11 @@ export class Request {
         return this._response(xhr, responseType);
     }
 
-    get<T>(url: string, responseType: ResponseTypes, headers?: Object): Observable<T> {
+    get<T>(url: string, responseType: ResponseTypes, headers?: Object, useCache?: boolean): Observable<T> {
```
Comment (Zlatkovsky) [inline src/client/app/services/request.ts]: generally an optional boolean parameter should be something where the default is `false`.  Here you have it as true.
Could you reverse the naming/functionality such that it's more of a `forceBypassCache`?

File: src/client/app/services/request.ts [hunk 14]
```diff
@@ -19,7 +19,11 @@ export class Request {
...
+        if (useCache != null && !useCache) {
+            url += (url.indexOf('?') === -1) ? '?' : '&';
```
Comment (Zlatkovsky) [inline src/client/app/services/request.ts]: How about a single statement?

            url = url + ((url.indexOf('?') === -1) ? '?' : '&') + `timestamp=${(new Date).getTime()}`;

File: src/client/app/actions/snippet.ts [hunk 15]
```diff
@@ -65,7 +65,7 @@ export class ImportSuccessAction implements Action {
 export class UpdateInfoAction implements Action {
     readonly type = SnippetActionTypes.UPDATE_INFO;
 
-    constructor(public payload: { name: string, description: string }) { }
+    constructor(public payload: { id: string, name?: string, description?: string, gist?: string, isOwned?: boolean }) { }
```
Comment (cshashwat) [inline src/client/app/actions/snippet.ts]: Discussed offline. This has been addressed by the field `gistOwnerId` in the `IGist` interface. The check is done in the actual app.

File: src/client/app/containers/app.ts [hunk 16]
```diff
@@ -18,8 +18,9 @@ import { Strings, environment } from '../helpers';
                     <command icon="OpenPaneMirrored" title="${Strings.runSideBySide}" (click)="runSideBySide()"></command>
                 </command>
                 <command [hidden]="isEmpty" icon="Share" [async]="sharing$|async" title="${Strings.share}">
-                    <command icon="PageCheckedin" title="${Strings.shareMenuPublic}" (click)="shareGist(true)"></command>
-                    <command icon="ProtectedDocument" title="${Strings.shareMenuPrivate}" (click)="shareGist(false)"></command>
+                    <command *ngIf="snippet?.isOwned" icon="Save" title="${Strings.updateMenu}" (click)="shareGist(null, true)"></command>
```
Comment (cshashwat) [inline src/client/app/containers/app.ts]: Haven't been able to find a way to leave inline comments in the Angular template. Perhaps HTML comments, or there's something I've missed? Would welcome any suggestions.

File: src/client/app/effects/github.ts [hunk 17]
```diff
@@ -84,10 +84,13 @@ export class GitHubEffects {
 
     @Effect()
     shareGist$: Observable<Action> = this.actions$
-        .ofType(GitHub.GitHubActionTypes.SHARE_PRIVATE_GIST, GitHub.GitHubActionTypes.SHARE_PUBLIC_GIST)
+        .ofType(GitHub.GitHubActionTypes.SHARE_PRIVATE_GIST,
+        GitHub.GitHubActionTypes.SHARE_PUBLIC_GIST,
+        GitHub.GitHubActionTypes.UPDATE_GIST)
```
Comment (cshashwat) [inline src/client/app/effects/github.ts]: Fixed.

File: src/client/app/effects/github.ts [hunk 18]
```diff
@@ -98,10 +101,14 @@ export class GitHubEffects {
             description.replace(Strings.gistDescriptionAppendage, ''); // shouldn't be necessary
             description += Strings.gistDescriptionAppendage;
 
+            if (type === GitHub.GitHubActionTypes.UPDATE_GIST && payload.isOwned) {
+                id = payload.gist;
```
Comment (cshashwat) [inline src/client/app/effects/github.ts]: Fixed.

File: src/client/app/effects/github.ts [hunk 19]
```diff
@@ -120,14 +127,21 @@ ${Strings.gistSharedDialogEnd}
...
-        .mergeMap(gist => Observable.from([
-            new GitHub.LoadGistsAction(),
-            new GitHub.ShareSuccessAction(gist)
-        ]))
-        .catch(exception => Observable.from([
-            new UI.ReportErrorAction(Strings.gistShareFailed, exception),
-            new GitHub.ShareFailedAction()
-        ]));
+        .mergeMap((gist) => Observable.from([
+                new GitHub.LoadGistsAction(),
```
Comment (cshashwat) [inline src/client/app/effects/github.ts]: Fixed.

File: src/client/app/effects/github.ts [hunk 20]
```diff
@@ -120,14 +127,21 @@ ${Strings.gistSharedDialogEnd}
...
+        )
+        .catch(exception => {
+            this._uiEffects.alert(
+                Strings.gistShareFailedBody + '\n\n' + Strings.reloadPrompt,
+                Strings.gistShareFailedTitle,
+                Strings.okButtonLabel)
+            .then(() => window.location.reload());
+            return Observable.from([
+                new GitHub.ShareFailedAction(),
+                new Snippet.UpdateInfoAction({ id: storage.lastOpened.id, gist: '', isOwned: false })]);
```
Comment (cshashwat) [inline src/client/app/effects/github.ts]: Thanks for bringing this up. Have modified code to pass through the `id` from the `payload` up above. That should address concerns with using `storage` (which has now been removed) and the un-owning of the snippet (since the `isOwned` field has now also been removed).

File: src/client/app/effects/snippet.ts [hunk 21]
```diff
@@ -144,6 +144,37 @@ export class SnippetEffects {
...
+    updateInfo$: Observable<Action> = this.actions$
+        .ofType(Snippet.SnippetActionTypes.UPDATE_INFO)
+        .map(( { payload } ) => {
+            let { id, name, description, gist, isOwned } = payload;
+            let snippet: ISnippet = storage.lastOpened;
+            if (storage.snippets.contains(id)) {
+                snippet = storage.snippets.get(id);
+
+                /* check if fields are undefined or null */
+                if (name != null) {
```
Comment (cshashwat) [inline src/client/app/effects/snippet.ts]: Discussed offline. Have fixed this by adding the `isNil()` from lodash.

File: src/client/app/effects/snippet.ts [hunk 22]
```diff
@@ -144,6 +144,37 @@ export class SnippetEffects {
...
+                    snippet.name = name;
+                }
+                if (description != null) {
```
Comment (cshashwat) [inline src/client/app/effects/snippet.ts]: Addressed above.

File: src/client/app/effects/snippet.ts [hunk 23]
```diff
@@ -144,6 +144,37 @@ export class SnippetEffects {
...
+                if (gist != null) {
+                    snippet.gist = gist;
+                }
+                if (isOwned != null) {
+                    snippet.isOwned = isOwned;
+                }
+
+                storage.snippets.insert(id, snippet);
+            }
+            return snippet;
```
Comment (cshashwat) [inline src/client/app/effects/snippet.ts]: Fixed.

File: src/client/app/helpers/snippet.helper.ts [hunk 24]
```diff
@@ -29,6 +29,7 @@ const snippetFields: { [key: string]: SnippetFieldType } = {
     /* ITemplate base class */
     id: SnippetFieldType.INTERNAL,
     gist: SnippetFieldType.INTERNAL,
+    isOwned: SnippetFieldType.INTERNAL,
```
Comment (cshashwat) [inline src/client/app/helpers/snippet.helper.ts]: Have addressed this by adding your suggestion. Thanks.

File: src/client/app/helpers/snippet.helper.ts (hunk 10 above)
Comment (cshashwat) [inline src/client/app/helpers/snippet.helper.ts]: Had forgotten to revert this change I'd made while I was testing out some code. Thanks for pointing it out. Fixed.

File: src/client/app/helpers/strings.ts (hunk 11 above)
Comment (cshashwat) [inline src/client/app/helpers/strings.ts]: Modified strings and added `/* NEEDS STRING REVIEW */` for all strings I've added to the file.

File: src/client/app/containers/app.ts [hunk 25]
```diff
@@ -64,6 +66,18 @@ export class AppComponent {
         return /commands=1/ig.test(location.search);
     }
 
+    get isOwned() {
```
Comment (Zlatkovsky) [inline src/client/app/containers/app.ts]: I still think we want a better name for it.

File: src/client/app/containers/app.ts [hunk 26]
```diff
@@ -18,8 +19,9 @@ import { Strings, environment } from '../helpers';
                     <command icon="OpenPaneMirrored" title="${Strings.runSideBySide}" (click)="runSideBySide()"></command>
                 </command>
                 <command [hidden]="isEmpty" icon="Share" [async]="sharing$|async" title="${Strings.share}">
-                    <command icon="PageCheckedin" title="${Strings.shareMenuPublic}" (click)="shareGist(true)"></command>
-                    <command icon="ProtectedDocument" title="${Strings.shareMenuPrivate}" (click)="shareGist(false)"></command>
+                    <command *ngIf="isOwned|async" icon="Save" title="${Strings.updateMenu}" (click)="shareGist(false, true)"></command>
```
Comment (Zlatkovsky) [inline src/client/app/containers/app.ts]: please add comments to boolean (or any other non-obvious) paramaters.

e.g.,

    shareGist(false /*parameter1Name*/, true /*parameter2Name*/)

File: src/client/app/effects/github.ts [hunk 27]
```diff
@@ -98,14 +103,20 @@ export class GitHubEffects {
             description.replace(Strings.gistDescriptionAppendage, ''); // shouldn't be necessary
             description += Strings.gistDescriptionAppendage;
 
+            if (type === GitHub.GitHubActionTypes.UPDATE_GIST) {
+                gistId = gist;
+            }
+
             return this._github.createOrUpdateGist(
                 `${description}`,
```
Comment (Zlatkovsky) [inline src/client/app/effects/github.ts]: isn't that just the same as `description`?

File: src/client/app/effects/github.ts [hunk 28]
```diff
@@ -98,14 +103,20 @@ export class GitHubEffects {
...
                 files,
-                null,
-                type === GitHub.GitHubActionTypes.SHARE_PUBLIC_GIST
+                gistId,
+                type === GitHub.GitHubActionTypes.SHARE_PUBLIC_GIST).map((gist: IGist) => {
```
Comment (Zlatkovsky) [inline src/client/app/effects/github.ts]: could we make this whole thing read cleaner?

    let isPublicGist = type === GitHub.GitHubActionTypes.SHARE_PUBLIC_GIST
    return this._github
        .createOrUpdateGist(
            description, files, gistId, isPublicGist
        )
        .map(gist: IGist => ({ gist: gist, snippetId: id} ));

File: src/client/app/effects/github.ts [hunk 29]
```diff
@@ -98,14 +103,20 @@ export class GitHubEffects {
...
+                    return { gist: gist, snippetId: id };
```
Comment (Zlatkovsky) [inline src/client/app/effects/github.ts]: if you wrap it in parentheses (see my prettified code above), you don't need the separate

    {
        return { <something> };
    }
